sudoku_solver.h

```diff
@@ -0,0 +1,38 @@
+#pragma once
+#include "sudoku_board.h"
+#include <optional>
+#include <tuple>
+
+/**
+ * a tools class which can be used to solve a sudoku board.
+ */
+class SudokuSolver {
+public:
+  /**
+   * gets a board and fills inplace the empty cells with values, so the board
+   * will be in a WIN state. if its possible, returns the board.
+   * if the board doesnt have a solution, returns an empty optional object.
+   */
+  static std::optional<SudokuBoard> solve_board(SudokuBoard &board);
```

Comment:
You get the board by reference in order to solve it inplace, this is great.
On the otherhand, when you are done you return an optional of `SudokuBoard`, but optional cannot contain a reference, so it copies the board and then returns it.
So why did you solve it inplace on the first hand?